### ISSUE TITLE
bucket_ref on replica does not prevent bucket move

### ISSUE THREAD

**Issue by Gerold103** [tarantool/vshard#173]
Bucket refs are used to prevent bucket movements during an active request. But only master now makes a decision about rebalancing, without consulting replicas. Master should ask replicas if they have any refs.

Maybe it will require prohibition of rebalancing when a replica is not available.

**Comment by Gerold103** [comment]
Solution for ro_refs: bucket GC should not delete bucket content until all replicas confirmed they have no ro_refs. This does not block rebalancing or anything. GC will just be postponed a little. This will work for current vshard since it has no master-master support.

For #165 rw_refs should be checked as well on all masters. And it is significantly more complex. A master, decided to move a bucket, should not start, until the bucket is readonly on all other masters. Problems emerge from 1) unavailability of some other masters, 2) a necessity to make a bucket read-write back after some timeout.

Maybe it will require modification of _bucket format to make communications about refs easier. Refs will be hold in lua tables still, but decisions about making a bucket read-write/read-only/gc will be done via special fields in _bucket tuples.

**Comment by R-omk** [comment]
Another solution:
before starting the migration of the bucket, wait until the completion of all requests on all replicas except for the replica on which the rebalancer worker is running.
At this time, all new requests should be redirected by the router to the main replica.
Now it can proceed with the migration of the bucket and enable rw locks.

**Comment by R-omk** [comment]
The replica could use a bucket crowding out algorithm similar to the algorithm that could be used here
https://github.com/tarantool/vshard/issues/219

**Comment by Gerold103** [comment]
Another raw idea: let replicas work independently and keep their refs. Let masters send buckets like now. But!!! on the replica side subscribe on _bucket changes. When applier on the replica will receive from the master node a request to change bucket state to sending/receiving/garbage, don't apply that request until all refs are gone. Don't know how to do that yet though. Because applier is async, and because can't yield for memtx spaces (in on_replace triggers). However maybe this idea could be elaborated. For instance, raise an error in the applier when master tries to change state of a bucket which has incompatible refs on this replica. This will break replication, but at least will prevent the existing requests from reading garbage.

**Comment by Gerold103** [comment]
There is another cool idea from @R-omk which I like the most among the listed ones. The idea is basically in "let it crash", but instead "let it fail".

The ticket is about RO refs, because until #165 the master is just one. And I am not sure #165 will be implemented ever anyway. With horizontal scaling you don't need multiple masters in a replicaset.

With RO refs the worst thing which can happen is that a request will read data from a bucket which is moved or deleted from the replica. Reading bad data is fine if in the end you can tell that it is bad and you need to retry. Nothing critical affecting other requests or data consistency happens.

Here is the punchline: when replica gets changes from master via replication about a bucket contradicting with local RO refs of this bucket - invalidate the ref. Drop it to zero or delete the ref entirely.

`storage_call()` in the end will try to make `bucket_unref()` and will fail because the ref was invalidated. And the request will return an error. Error can have some special code supported by router which would led to automatic retry.
